### PR TITLE
AI loses vision when cameras lose power.

### PR DESCRIPTION
**Pull request by Mervill**
## About the PR

The AI will now properly lose vision granted by cameras when those cameras lose power.

## Media

ai-power

## Requirements

- [x] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- fix: The AI will now properly lose vision granted by cameras when those cameras lose power

### CODE REVIEW

**Comment by superjj18** [comment]
Perhaps a happy compromise would be giving cameras a small internal battery similar to flashing alert lights?  Gives a buffer of time before the ai gets completely blinded station wide

**Comment by baddiePRO** [comment]
do EMPs affect cameras? I haven't messed with them enough to know

**Comment by Mervill** [comment]
Yes EMP does affect cameras

**Comment by PJB3005** [comment]
> Perhaps a happy compromise would be giving cameras a small internal battery similar to flashing alert lights? Gives a buffer of time before the ai gets completely blinded station wide

I'd rather just have APC channels implemented instead. I don't want us to get into the habit of adding ever more batteries to machinery around the station.

**Comment by Minigold07** [comment]
Maybe just reduce the vision or something? If a full power outtage happens when nukies attack for example the AI is blind for the rest of the round. I don't think it would be fun to be unable to see anything because of minor power issues or a singulo going off.

**Comment by ImWeax** [comment]
The AI should be able to see if there is a disabled camera in the room, maybe highlight the camera in red over the AI overlay?

**Comment by ScarKy0** [comment]
How about instead this will be like an optional thing in AiVisionComponent?
Like requiresPower, changeWhenUnpowered and rangeWhenUnpowered or something?
Could also be put on other things that way

**Comment by Mervill** [comment]
Right now it only applies to `AiVisionComponent`s that are attached to Surveillance Cameras. The vision component that is (for example) attached to the AI core isn't affected by this.

**Comment by ScarKy0** [comment]
> Right now it only applies to AiVisionComponents that are attached to Surveillance Cameras. The vision component that is (for example) attached to the AI core isn't affected by this.

Yeah, im just putting this down as a possible idea

**Comment by Psychpsyo** [comment]
Wasn't there talk about the AI being able to see through borgs as well?
That would help with the "no power = no vision" problem, depending on how it gets implemented.

**Comment by nymhole** [comment]
ai needs a way to move its camera and know where it is without vision, and also should not HAVE to see something to do something, or know that something its connected to is there.

**Comment by nymhole** [comment]
> How about instead this will be like an optional thing in AiVisionComponent? Like requiresPower, changeWhenUnpowered and rangeWhenUnpowered or something? Could also be put on other things that way
this is a great idea for versatility, admeme events would be very happy too im sure.

**Comment by Emisse** [comment]
could just change it to not be able to interact with things in dead cams range if people dont want loss of vision and just tie vision to decon/wire cutting or something

**Comment by ArcaneOcto** [comment]
Just an idea: what if instead of completely shrouding a room with static, it could be partially covered. The static system appears to tile based, so maybe it could randomly tiles to cover with static? Maybe like 70% reduced visibilty? Which tiles are concealed could be random and refresh every couple seconds. Thus making the AI have to watch the unpowered camera for longer or risk missing something.

File: Content.Server/Silicons/StationAi/StationAiSystem.cs [hunk 1]
```diff
@@ -73,4 +80,17 @@ private void AnnounceSnip(EntityUid entity)
         _chats.ChatMessageToMany(ChatChannel.Notifications, msg, msg, entity, false, true, filter.Recipients.Select(o => o.Channel));
         // Apparently there's no sound for this.
     }
+
+    private void OnSurveillanceCameraActivate(EntityUid camera, SurveillanceCameraComponent component, ref SurveillanceCameraActivateEvent args)
```
Comment (TheShuEd) [inline Content.Server/Silicons/StationAi/StationAiSystem.cs:84]: Use Entity<>

File: Content.Server/Silicons/StationAi/StationAiSystem.cs [hunk 2]
```diff
@@ -73,4 +80,17 @@ private void AnnounceSnip(EntityUid entity)
...
+    {
+        if (TryComp(args.Camera, out StationAiVisionComponent? aiVision))
+            SetVisionEnabled((camera, aiVision), true);
+    }
+
+    private void OnSurveillanceCameraDeactivate(EntityUid camera, SurveillanceCameraComponent component, ref SurveillanceCameraDeactivateEvent args)
```
Comment (TheShuEd) [inline Content.Server/Silicons/StationAi/StationAiSystem.cs:90]: Use Entity<>

File: Content.Server/Silicons/StationAi/StationAiSystem.cs (hunk 1 above)
Comment (TheShuEd) [reply Content.Server/Silicons/StationAi/StationAiSystem.cs:84]:
```suggestion
    private void OnSurveillanceCameraActivate(Entity<SurveillanceCameraComponent> camera, ref SurveillanceCameraActivateEvent args)
```
for example

you can get component by camera.Comp

File: Content.Server/SurveillanceCamera/Systems/SurveillanceCameraSystem.cs [hunk 3]
```diff
@@ -428,16 +433,11 @@ public sealed class OnSurveillanceCameraViewerRemoveEvent : EntityEventArgs
 
 }
 
-// What happens when a camera deactivates.
-public sealed class SurveillanceCameraDeactivateEvent : EntityEventArgs
-{
-    public EntityUid Camera { get; }
+[ByRefEvent]
+public record struct SurveillanceCameraActivateEvent(EntityUid Camera);
```
Comment (TheShuEd) [inline Content.Server/SurveillanceCamera/Systems/SurveillanceCameraSystem.cs:437]: you can combine these events into a single event, with the new camera status information being transmitted

Comment (Mervill) [reply Content.Server/SurveillanceCamera/Systems/SurveillanceCameraSystem.cs:437]: I'm always on the fence about when to do this (using a bool vs defining two events). Beyond style choices and mild logic changes, is there a reason to favor one over the other?

Comment (TheShuEd) [reply Content.Server/SurveillanceCamera/Systems/SurveillanceCameraSystem.cs:437]: Maybe not.
You can, not should

**Comment by metalgearsloth** [comment]
We discussed this among maintainers and at the moment it's not really good for gameplay because if there's a power outage, the AI can't see or do anything and just sits there with no agency (e.g. powergridcheck, station loses power over time). If there were navmap overlay or borg vision it might be more interesting but would need to reassess then.

**Pull request closed without merging**